IPython/core/interactiveshell.py

```diff
@@ -2682,14 +2682,18 @@ def error_before_exec(value):
                 # Compile to bytecode
                 try:
                     code_ast = compiler.ast_parse(cell, filename=cell_name)
+                except self.custom_exceptions as e:
+                    etype, value, tb = sys.exc_info()
+                    self.CustomTB(etype, value, tb)
+                    return error_before_exec(e)
                 except IndentationError as e:
                     self.showindentationerror()
                     if store_history:
                         self.execution_count += 1
                     return error_before_exec(e)
                 except (OverflowError, SyntaxError, ValueError, TypeError,
                         MemoryError) as e:
-                    self.showsyntaxerror()
```

Comment:
Yes it should, but it seem hard to follow the logic if some places call `syntaxerror()` and others `showtraceback()`, 
so I tried to simplify by onl calling `showtraceback()`, still I think this might be wrong. I'll revert.